src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php

```diff
@@ -13,5 +13,5 @@ abstract class AbstractCartItemAddRoute
 {
     abstract public function getDecorated(): AbstractCartItemAddRoute;
 
-    abstract public function add(Request $request, Cart $cart, SalesChannelContext $context, ?array $items): CartResponse;
+    abstract public function add(Request $request, Cart $cart, SalesChannelContext $context, ?array $items, bool $dispatchAfter): CartResponse;
```

Comment:
why did you add this parameter everywhere? what is the purpose of this?

Reply:
If you subscribe to the event `AfterLineItemAddedEvent` and the subscriber is adding a gift product after a specific product is added to the cart it needs a flag to prevent the event being fired again or it ends in an endless loop.

Reply:
to prevent this, you should add an extension to the `Context` object and check for this extension in your subscriber

besides that, adding a new required parameter to a method is a breaking change :wink:

Reply:
The parameter has a default value to make it optional. My subscriber to add the gift is using the CartService to add it to the cart and by using this class it goes over the CartItemAddRoute, so there is no way to prevent it in a subscriber, it has to be done in the CartItemAddRoute.

Reply:
you said, you are in an endless loop because you subscribe to the same event which would be fired in your subscriber could.
1. this would affect all events and not only the `After...` events :wink: 
2. you could simply do the following:
    - event subscriber method is called 
    - get the `\Shopware\Core\Framework\Context` from the given *Event object 
    - add an extension to the context object (`\Shopware\Core\Framework\Struct\ExtendableTrait::addExtension`)
    - add your new item to the cart
    - event subscriber method is called again, because you add an item to the cart
    - check if the Context object contains your extension. if so, remove the extension from the context, return and do nothing. if it does not contain your extension, your code will be executed normally

Reply:
Ah now I see it :D let me try it quickly, then I'll remove the parameter

Reply:
I've made the changes, but I noticed If anyone has multiple plugins subscribing to the AfterLineItemAddedEvent and there is one plugin doing this add gift, all of these subscribers needs to check if the context has the extension and return the function. So by using the parameter to not dispatch the events seems better to me, but If you insist to have it like that, I leave it to this current state.

Reply:
Yes that's fine, any subscriber will need to take these precautions if it works similarly to yours. Those other subscribers would have their own context extension to check for.